crypto/evp/m_sha3.c

```diff
@@ -123,12 +139,88 @@ static int sha3_final(EVP_MD_CTX *evp_ctx, unsigned char *md)
     ctx->buf[bsz - 1] |= 0x80;
 
     (void)SHA3_absorb(ctx->A, ctx->buf, bsz, bsz);
+}
+
+static int sha3_final(EVP_MD_CTX *evp_ctx, unsigned char *md)
+{
+    KECCAK1600_CTX *ctx = evp_ctx->md_data;
+    size_t bsz = ctx->block_size;
+
+    sha3_final_absorb(evp_ctx);
 
     SHA3_squeeze(ctx->A, md, ctx->md_size, bsz);
 
     return 1;
 }
 
+static int shake_final(EVP_MD_CTX *evp_ctx, unsigned char *md)
+{
+    KECCAK1600_CTX *ctx = evp_ctx->md_data;
+    size_t bsz = ctx->block_size;
+    size_t md_size = ctx->md_size;
+    size_t num = ctx->num;
+
+    /* We may need to invoke KeccakF directly, since SHA3_squeeze does not
+     * invoke it itself at the end of a block.
+     */
+    int need_keccak = (ctx->squeezing == 2);
+
+    if (! ctx->squeezing) {
+        sha3_final_absorb(evp_ctx);
+        ctx->squeezing = 1;
+        num = ctx->num = 0;
+        need_keccak = 0; /* "absorb" runs the round function when done. */
+        OPENSSL_cleanse(ctx->buf, sizeof(ctx->buf));
+    }
+
+    if (!md_size) {
+        return 1;
+    }
+
+    if (num) {
+        /* We have some cached data in ctx->buf; we should serve it. */
+        if (md_size <= num) {
+            /* We have enough cached data in buf; we can just copy it out. */
+            memcpy(md, ctx->buf + (bsz-num), md_size);
+            OPENSSL_cleanse(ctx->buf+(bsz-num), md_size);
+            ctx->num -= md_size;
+            return 1;
+        }
+        memcpy(md, ctx->buf + (bsz-num), num);
+        OPENSSL_cleanse(ctx->buf+(bsz-num), num);
+        md_size -= num;
+        md += num;
+        ctx->num = 0;
+    }
+
+    ctx->squeezing = 2;
+
+    if (need_keccak) {
+        KeccakF1600(ctx->A);
+    }
+
+    /* These bytes will be part of a partial block, copied from buf. */
+    size_t remainder = md_size % bsz;
+    /* We can write this many bytes to md directly. */
+    size_t direct_bytes = md_size - remainder;
```

Comment:
Sorry, we need to abide to C90 syntax, which means that variable definitions at the beginning of blocks only.

There are two ways to do that.  One is to move the variable definitions (but not the assignments) to the top of the functions, the other is to form an unconditional block that wraps the use of these variables.